src/ccl_cls.c

```diff
@@ -0,0 +1,1336 @@
+      free(clt);
+      *status=CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,"ccl_cls.c: ccl_cl_tracer(): memory allocation\n");
+      return NULL;
+    }
+
+    gsl_integration_workspace *w=gsl_integration_workspace_alloc(ccl_gsl->N_ITERATION);
+    F.function=&speval_bis;
+    F.params=clt->spl_nz;
+    gslstatus=gsl_integration_qag(&F, z_n[0], z_n[nz_n-1], 0,
+                                  ccl_gsl->INTEGRATION_EPSREL, ccl_gsl->N_ITERATION,
+                                  ccl_gsl->INTEGRATION_GAUSS_KRONROD_POINTS,
+                                  w, &nz_norm, &nz_enorm);
+    gsl_integration_workspace_free(w);
+    if(gslstatus!=GSL_SUCCESS) {
+      ccl_raise_gsl_warning(gslstatus, "ccl_cls.c: cl_tracer():");
+      ccl_spline_free(clt->spl_nz);
+      free(clt);
+      *status=CCL_ERROR_INTEG;
+      strcpy(cosmo->status_message,"ccl_cls.c: ccl_cl_tracer(): integration error when normalizing N(z)\n");
+      return NULL;
+    }
+    for(int ii=0;ii<nz_n;ii++)
+      nz_normalized[ii]=n[ii]/nz_norm;
+    ccl_spline_free(clt->spl_nz);
+    clt->spl_nz=ccl_spline_init(nz_n,z_n,nz_normalized,0,0);
+    free(nz_normalized);
+    if(clt->spl_nz==NULL) {
+      free(clt);
+      *status=CCL_ERROR_SPLINE;
+      strcpy(cosmo->status_message,"ccl_cls.c: ccl_cl_tracer(): error initializing normalized spline for N(z)\n");
+      return NULL;
+    }
+
+    if(tracer_type==CL_TRACER_NC) {
+      //Initialize bias spline
+      clt->spl_bz=ccl_spline_init(nz_b,z_b,b,b[0],b[nz_b-1]);
+      if(clt->spl_bz==NULL) {
+	ccl_spline_free(clt->spl_nz);
+	free(clt);
+	*status=CCL_ERROR_SPLINE;
+	strcpy(cosmo->status_message,"ccl_cls.c: ccl_cl_tracer(): error initializing spline for b(z)\n");
+	return NULL;
+      }
+      clt->has_rsd=has_rsd;
+      clt->has_magnification=has_magnification;
+      if(clt->has_magnification) {
+	//Compute weak lensing kernel
+	int nchi;
+	double *x,*y;
+	double dchi_here=5.;
+	double zmax=clt->spl_nz->xf;
+	double chimax=ccl_comoving_radial_distance(cosmo,1./(1+zmax),status);
+	//TODO: The interval in chi (5. Mpc) should be made a macro
+
+	//In this case we need to integrate all the way to z=0. Reset zmin and chimin
+	clt->zmin=0;
+	clt->chimin=0;
+	clt->spl_sz=ccl_spline_init(nz_s,z_s,s,s[0],s[nz_s-1]);
+	if(clt->spl_sz==NULL) {
+	  ccl_spline_free(clt->spl_nz);
+	  ccl_spline_free(clt->spl_bz);
+	  free(clt);
+	  *status=CCL_ERROR_SPLINE;
+	  strcpy(cosmo->status_message,
+		 "ccl_cls.c: ccl_cl_tracer(): error initializing spline for s(z)\n");
+	  return NULL;
+	}
+
+	nchi=(int)(chimax/dchi_here)+1;
+	x=ccl_linear_spacing(0.,chimax,nchi);
+	dchi_here=chimax/nchi;
+	if(x==NULL || (fabs(x[0]-0)>1E-5) || (fabs(x[nchi-1]-chimax)>1e-5)) {
+	  ccl_spline_free(clt->spl_nz);
+	  ccl_spline_free(clt->spl_bz);
+	  ccl_spline_free(clt->spl_sz);
+	  free(clt);
+	  *status=CCL_ERROR_LINSPACE;
+	  strcpy(cosmo->status_message,
+		 "ccl_cls.c: ccl_cl_tracer(): Error creating linear spacing in chi\n");
+	  return NULL;
+	}
```

Comment:
@rmjarvis just to clarify: this can be done at the expense of enclosing a large fraction of the code in `if(*status==0)` clauses. Is this what you suggest?

Reply:
Right.  If you want to see how that looks, @villarrealas did this for the `ccl_cosmology_compute_sigma` function [here](https://github.com/LSSTDESC/CCL/blob/master/src/ccl_massfunc.c#L431).  I think it's cleaner this way, since the frees only show up in one place, as there are no early returns.  So less likely to make a mistake and miss getting the logic right about what needs to be freed when.

Reply:
Addressed in #506